src/goto-symex/symex_goto.cpp

```diff
@@ -65,6 +67,182 @@ void goto_symext::apply_goto_condition(
+  const constant_exprt *constant_expr =
+    expr_try_dynamic_cast<constant_exprt>(other_operand);
+
+  INVARIANT(
+    other_operand.id() == ID_address_of ||
+      (typecast_expr && typecast_expr->op().id() == ID_address_of) ||
+      (constant_expr && constant_expr->get_value() == ID_NULL),
+    "An expression that passes goto_symex_is_constant and has "
+    "pointer-type must be an address of expression or a null pointer");
+
+  const ssa_exprt *ssa_symbol_expr =
+    expr_try_dynamic_cast<ssa_exprt>(symbol_expr);
+
+  value_setst::valuest value_set_elements;
+  value_set.get_value_set(
+    ssa_symbol_expr->get_l1_object(), value_set_elements, ns);
+
+  bool constant_found = false;
+
+  for(const auto &value_set_element : value_set_elements)
+  {
+    if(
+      value_set_element.id() == ID_unknown ||
+      value_set_element.id() == ID_invalid)
+    {
+      // If ID_unknown or ID_invalid are in the value-set then we can't
+      // conclude anything about it
+      return;
+    }
+
+    if(!constant_found)
+    {
+      optionalt<exprt> reference =
+        value_set_dereferencet::build_reference_to_value_set_element(
+          value_set_element,
+          symbol_expr,
+          to_pointer_type(symbol_expr.type()),
+          language_mode,
+          ns);
+
+      if(reference && *reference == other_operand)
+      {
+        constant_found = true;
+        // We can't break because we have to make sure we find any instances of
+        // ID_unknown or ID_invalid
+      }
+    }
+  }
+
+  if(!constant_found)
+  {
+    // The symbol cannot possibly have the value \p other_operand because it
+    // isn't in the symbol's value-set
+    condition = condition.get().id() == ID_equal ? make_renamed<L2>(false_exprt{})
+                                           : make_renamed<L2>(true_exprt{});
+  }
+  else if(value_set_elements.size() == 1)
+  {
+    // The symbol must have the value \p other_operand because it is the only
+    // thing in the symbol's value-set
+    condition = condition.get().id() == ID_equal ? make_renamed<L2>(true_exprt{})
+                                           : make_renamed<L2>(false_exprt{});
+  }
+}
+
+
+
+/// Check if we have a simple pointer comparison, and if so try to evaluate it.
+/// \param [in,out] condition: An L2- renamed expression of the form
+///   "symbol_expr == other" or "symbol_expr != other" (or one of those with
+///   the lhs and rhs swapped)
+/// \param symbol_expr: The symbol expression in the condition
+/// \param other_operand: The other expression in the condition - must pass
+///   goto_symex_is_constant, and since it is pointer-typed it must therefore
+///   be an address of expression, a typecast of an address of expression or a
+///   null pointer
+/// \param value_set: The value-set for looking up what the symbol can point to
+/// \param language_mode: The language mode
+/// \param ns: A namespace
+static bool try_evaluate_pointer_comparisons_base_case_with_check(
```

Comment:
Instead of calling this twice, suggest something like https://github.com/diffblue/cbmc/blob/develop/src/goto-symex/goto_state.cpp#L67